File: shared/go_libs/validation/network.go
package validation

import (
	"net"
	"net/url"
	"regexp"
	"strings"
)

// Email validates that a string is a valid email address.
func Email(opts ...func(*emailOpts)) Validator {
	o := &emailOpts{
		normalize:    true,
		errorMessage: "invalid email address",
	}
	for _, opt := range opts {
		opt(o)
	}

	// RFC 5322 compliant email regex (simplified but robust)
	emailRegex := regexp.MustCompile(
		`^[a-zA-Z0-9.!#$%&'*+/=?^_` + "`" + `{|}~-]+` +
			`@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?` +
			`(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$`,
	)

	return &StringValidator{
		validate: func(s string) ValidationResult {
			email := strings.TrimSpace(s)
			if email == "" {
				return Failure(o.errorMessage)
			}

			// Check length constraints
			if len(email) > 254 {
				return Failure(o.errorMessage)
			}

			// Check local part length
			atIdx := strings.Index(email, "@")
			if atIdx == -1 || atIdx > 64 {
				return Failure(o.errorMessage)
			}

			if !emailRegex.MatchString(email) {
				return Failure(o.errorMessage)
			}

			if o.normalize {
				email = strings.ToLower(email)
			}

			return Success(email)
		},
	}
}

type emailOpts struct {
	normalize    bool
	errorMessage string
}

// WithoutNormalize disables email normalization (lowercase).
func WithoutNormalize() func(*emailOpts) {
	return func(o *emailOpts) {
		o.normalize = false
	}
}

// WithEmailError sets a custom error message for Email.
func WithEmailError(msg string) func(*emailOpts) {
	return func(o *emailOpts) {
		o.errorMessage = msg
	}
}

// URL validates that a string is a valid URL.
func URL(opts ...func(*urlOpts)) Validator {
	o := &urlOpts{
		requireTLD:     true,
		allowedSchemes: map[string]struct{}{"http": {}, "https": {}},
		errorMessage:   "invalid URL",
	}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			urlStr := strings.TrimSpace(s)
			if urlStr == "" {
				return Failure(o.errorMessage)
			}

			parsed, err := url.Parse(urlStr)
			if err != nil {
				return Failure(o.errorMessage)
			}

			// Check scheme
			if parsed.Scheme == "" {
				return Failure(o.errorMessage)
			}

			if _, ok := o.allowedSchemes[strings.ToLower(parsed.Scheme)]; !ok {
				schemes := make([]string, 0, len(o.allowedSchemes))
				for s := range o.allowedSchemes {
					schemes = append(schemes, s)
				}
				return Failuref("URL scheme must be one of: %s", strings.Join(schemes, ", "))
			}

			// Check netloc (hostname)
			if parsed.Host == "" {
				return Failure(o.errorMessage)
			}

			// Check for TLD if required
			if o.requireTLD {
				hostname := parsed.Hostname()
				if !strings.Contains(hostname, ".") && strings.ToLower(hostname) != "localhost" {
					return Failure(o.errorMessage)
				}
			}

			return Success(urlStr)
		},
	}
}

type urlOpts struct {
	requireTLD     bool
	allowedSchemes map[string]struct{}
	errorMessage   string
}

// WithoutTLD allows URLs without a top-level domain.
func WithoutTLD() func(*urlOpts) {
	return func(o *urlOpts) {
		o.requireTLD = false
	}
}

// WithSchemes sets allowed URL schemes.
func WithSchemes(schemes ...string) func(*urlOpts) {
	return func(o *urlOpts) {
		o.allowedSchemes = make(map[string]struct{}, len(schemes))
		for _, s := range schemes {
			o.allowedSchemes[strings.ToLower(s)] = struct{}{}
		}
	}
}

// WithURLError sets a custom error message for URL.
func WithURLError(msg string) func(*urlOpts) {
	return func(o *urlOpts) {
		o.errorMessage = msg
	}
}

// IPAddress validates that a string is a valid IP address.
func IPAddress(opts ...func(*ipOpts)) Validator {
	o := &ipOpts{
		version: 0, // 0 = any, 4 = IPv4, 6 = IPv6
	}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			ipStr := strings.TrimSpace(s)
			if ipStr == "" {
				return Failure(o.getErrorMessage())
			}

			ip := net.ParseIP(ipStr)
			if ip == nil {
				return Failure(o.getErrorMessage())
			}

			// Check version if specified
			isIPv4 := ip.To4() != nil
			if o.version == 4 && !isIPv4 {
				return Failure("value must be a valid IPv4 address")
			}
			if o.version == 6 && isIPv4 {
				return Failure("value must be a valid IPv6 address")
			}

			return Success(ipStr)
		},
	}
}

type ipOpts struct {
	version      int // 0 = any, 4 = IPv4, 6 = IPv6
	errorMessage string
}

func (o *ipOpts) getErrorMessage() string {
	if o.errorMessage != "" {
		return o.errorMessage
	}
	switch o.version {
	case 4:
		return "value must be a valid IPv4 address"
	case 6:
		return "value must be a valid IPv6 address"
	default:
		return "value must be a valid IP address"
	}
}

// IPv4Only restricts IPAddress to IPv4 only.
func IPv4Only() func(*ipOpts) {
	return func(o *ipOpts) {
		o.version = 4
	}
}

// IPv6Only restricts IPAddress to IPv6 only.
func IPv6Only() func(*ipOpts) {
	return func(o *ipOpts) {
		o.version = 6
	}
}

// WithIPError sets a custom error message for IPAddress.
func WithIPError(msg string) func(*ipOpts) {
	return func(o *ipOpts) {
		o.errorMessage = msg
	}
}

// Hostname validates that a string is a valid hostname.
func Hostname(opts ...func(*hostnameOpts)) Validator {
	o := &hostnameOpts{
		allowIP:      false,
		requireTLD:   false,
		errorMessage: "invalid hostname",
	}
	for _, opt := range opts {
		opt(o)
	}

	// RFC 1123 hostname pattern
	hostnameRegex := regexp.MustCompile(
		`^(?i)[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?` +
			`(\.[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)*$`,
	)

	return &StringValidator{
		validate: func(s string) ValidationResult {
			hostname := strings.TrimSpace(strings.ToLower(s))
			if hostname == "" {
				return Failure(o.errorMessage)
			}

			// Check total length
			if len(hostname) > 253 {
				return Failure(o.errorMessage)
			}

			// Check if it's an IP address
			if o.allowIP {
				if ip := net.ParseIP(hostname); ip != nil {
					return Success(hostname)
				}
			}

			// Validate hostname format
			if !hostnameRegex.MatchString(hostname) {
				return Failure(o.errorMessage)
			}

			// Check TLD requirement
			if o.requireTLD && !strings.Contains(hostname, ".") {
				return Failure("hostname must have a top-level domain")
			}

			return Success(hostname)
		},
	}
}

type hostnameOpts struct {
	allowIP      bool
	requireTLD   bool
	errorMessage string
}

// AllowIP allows IP addresses as hostnames.
func AllowIP() func(*hostnameOpts) {
	return func(o *hostnameOpts) {
		o.allowIP = true
	}
}

// RequireTLD requires hostnames to have a TLD.
func RequireTLD() func(*hostnameOpts) {
	return func(o *hostnameOpts) {
		o.requireTLD = true
	}
}

// WithHostnameError sets a custom error message for Hostname.
func WithHostnameError(msg string) func(*hostnameOpts) {
	return func(o *hostnameOpts) {
		o.errorMessage = msg
	}
}

File: shared/node_libs/src/validation/network.ts
/**
 * Network validators for input validation (email, URL, IP, hostname).
 */

import { type ValidationResult, type Validator, success, failure } from './index.js';

export interface EmailOptions {
  normalize?: boolean;
  errorMessage?: string;
}

/**
 * Validates that a string is a valid email address.
 * Uses RFC 5322 compliant regex pattern.
 */
export function email(options: EmailOptions = {}): Validator<string, string> {
  const { normalize = true } = options;
  const errorMessage = options.errorMessage ?? 'Invalid email address';

  // RFC 5322 compliant email regex (simplified but robust)
  const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$/;

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const emailStr = value.trim();
    if (emailStr === '') {
      return failure(errorMessage);
    }

    // Check length constraints
    if (emailStr.length > 254) {
      return failure(errorMessage);
    }

    // Check local part length
    const atIndex = emailStr.indexOf('@');
    if (atIndex === -1 || atIndex > 64) {
      return failure(errorMessage);
    }

    if (!emailRegex.test(emailStr)) {
      return failure(errorMessage);
    }

    return success(normalize ? emailStr.toLowerCase() : emailStr);
  };
}

export interface UrlOptions {
  requireTld?: boolean;
  allowedSchemes?: string[];
  errorMessage?: string;
}

/**
 * Validates that a string is a valid URL.
 */
export function url(options: UrlOptions = {}): Validator<string, string> {
  const {
    requireTld = true,
    allowedSchemes = ['http', 'https'],
  } = options;
  const errorMessage = options.errorMessage ?? 'Invalid URL';

  const schemeSet = new Set(allowedSchemes.map(s => s.toLowerCase()));

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const urlStr = value.trim();
    if (urlStr === '') {
      return failure(errorMessage);
    }

    let parsed: URL;
    try {
      parsed = new URL(urlStr);
    } catch {
      return failure(errorMessage);
    }

    // Check scheme
    const scheme = parsed.protocol.replace(':', '').toLowerCase();
    if (!schemeSet.has(scheme)) {
      return failure(`URL scheme must be one of: ${allowedSchemes.join(', ')}`);
    }

    // Check hostname
    if (!parsed.hostname) {
      return failure(errorMessage);
    }

    // Check for TLD if required
    if (requireTld) {
      const hostname = parsed.hostname.toLowerCase();
      if (!hostname.includes('.') && hostname !== 'localhost') {
        return failure(errorMessage);
      }
    }

    return success(urlStr);
  };
}

export interface IpAddressOptions {
  version?: 4 | 6;
  errorMessage?: string;
}

/**
 * Validates that a string is a valid IP address.
 */
export function ipAddress(options: IpAddressOptions = {}): Validator<string, string> {
  const { version } = options;

  // IPv4 regex
  const ipv4Regex = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;

  // Simplified IPv6 regex (covers most common formats)
  const ipv6Regex = /^(?:(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,7}:|(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}|(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}|(?:[0-9a-fA-F]{1,4}:){1,3}(?::[0-9a-fA-F]{1,4}){1,4}|(?:[0-9a-fA-F]{1,4}:){1,2}(?::[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:(?::[0-9a-fA-F]{1,4}){1,6}|:(?:(?::[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(?::[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]+|::(?:ffff(?::0{1,4})?:)?(?:(?:25[0-5]|(?:2[0-4]|1?[0-9])?[0-9])\.){3}(?:25[0-5]|(?:2[0-4]|1?[0-9])?[0-9])|(?:[0-9a-fA-F]{1,4}:){1,4}:(?:(?:25[0-5]|(?:2[0-4]|1?[0-9])?[0-9])\.){3}(?:25[0-5]|(?:2[0-4]|1?[0-9])?[0-9]))$/;

  const getErrorMessage = (): string => {
    if (options.errorMessage) return options.errorMessage;
    if (version === 4) return 'Value must be a valid IPv4 address';
    if (version === 6) return 'Value must be a valid IPv6 address';
    return 'Value must be a valid IP address';
  };

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const ipStr = value.trim();
    if (ipStr === '') {
      return failure(getErrorMessage());
    }

    const isIPv4 = ipv4Regex.test(ipStr);
    const isIPv6 = ipv6Regex.test(ipStr);

    if (!isIPv4 && !isIPv6) {
      return failure(getErrorMessage());
    }

    if (version === 4 && !isIPv4) {
      return failure('Value must be a valid IPv4 address');
    }

    if (version === 6 && !isIPv6) {
      return failure('Value must be a valid IPv6 address');
    }

    return success(ipStr);
  };
}

export interface HostnameOptions {
  allowIp?: boolean;
  requireTld?: boolean;
  errorMessage?: string;
}

/**
 * Validates that a string is a valid hostname.
 */
export function hostname(options: HostnameOptions = {}): Validator<string, string> {
  const {
    allowIp = false,
    requireTld = false,
  } = options;
  const errorMessage = options.errorMessage ?? 'Invalid hostname';

  // RFC 1123 hostname pattern
  const hostnameRegex = /^(?![-])[a-zA-Z0-9-]{1,63}(?<![])(?:\.(?![]))[a-zA-Z0-9-]{1,63}(?<![]))*$/;

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const hostStr = value.trim().toLowerCase();
    if (hostStr === '') {
      return failure(errorMessage);
    }

    // Check total length
    if (hostStr.length > 253) {
      return failure(errorMessage);
    }

    // Check if it's an IP address
    if (allowIp) {
      const ipResult = ipAddress()(hostStr);
      if (ipResult.isValid) {
        return success(hostStr);
      }
    }

    // Validate hostname format
    // Simpler validation: each label 1-63 chars, alphanumeric and hyphens, no leading/trailing hyphens
    const labels = hostStr.split('.');
    for (const label of labels) {
      if (label.length === 0 || label.length > 63) {
        return failure(errorMessage);
      }
      if (label.startsWith('-') || label.endsWith('-')) {
        return failure(errorMessage);
      }
      if (!/^[a-z0-9-]+$/.test(label)) {
        return failure(errorMessage);
      }
    }

    // Check TLD requirement
    if (requireTld && !hostStr.includes('.')) {
      return failure('Hostname must have a top-level domain');
    }

    return success(hostStr);
  };
}
